docs/modules/ROOT/pages/db_services/alcor-database-interface.adoc

```diff
@@ -0,0 +1,146 @@
+
+The SqlFieldsQuery is executed using query() method of Ignite. It returns an instance of QueryCursor object representing the result set (list of rows and associated metadata. Iterating over the rows of the cursor a result set usable by the Alcor code is built and returned.
+
+An optimization not attempted at this is to compile these queries once at the time the cache is built, or an instance handle is obtained (called prepared statement in DBMS parlance) because the memory requirements could become very high. It is possible to enable this optimization selectively for the queries whose query compilation time dominates the execution time.
+
+==== ScanQuery
+When a query with queryParams (search using non key fields) is not eligible for SQL index query, it is executed using ScanQuery interface of Ignite. The search condition represented by the query params is used to construct an instance of igniteBiPredicate, which is used to build an instance of ScanQuery. The query() method of Ignite also works with ScanQuery and returns a QueryCursor.
+
+=== Transactions
+Transactions are applicable and required only on caches created with TRANSACTIONAL atomicity mode.
+In some places in the code, more than one cache is operated on and all these operations have to done or none of them should be done. This is where Alcor stack uses transactions. DB module uses start(), commit(), rollback() methods of Ignite to facilitate transactions.
+
+=== Distributed Locks
+Ignite transactions are implicitly distributed but in some cases explicit distributed locks may be required. Alcor stack uses Ignite cache to simulate distributed lock. Distributed locks are used in Alcor primarily for locking portions of the cache instead of the whole cache.
+
+This is an optimization aimed at increasing concurrency.
+
+Only ElasticIpAllocator (Elastic IP Manager), BitmapPoolImpl and MacServiceImpl (MAC Manager), and QuotaServiceImpl [QM] use distributed locks.
+
+=== Optimization features implemented
+. SQL Indices to speedup non key field lookup.
+. Store index field value of upto 36 characters inline to avoid multiple index page lookups.
+. Distributed locks to reduce lock contention and increase concurrency.
+. Enable Partition awareness to avoid one ignite server node to become bottleneck.
+. Use of bulk get/put operations.
+
+=== Optimization features to consider
+. Thick client with near caches.
+. Keeping data, checkpoint and WAL storage separate, preferably on SSD.
+. Adjusting WAL segment size.
+. Enabling Direct I/O.
+. Enabling Binary Mode for select caches.
+
+==== Summary of Key Ignite interfaces and features used in Alcor DB module
+. IgniteClient
+. ClientCacheConfiguration - with the following attributes:
+.. expiration policy
+.. custom cache name
+.. atomicity mode for transactions
+.. SQL Schema name
+.. QueryEntity
+.. QueryIndex
+
+. SqlFieldsQuery
+. QueryCursor
+. ScanCursor
+. IgniteBiPredicate
+. Transaction
+
```

Comment:
As discussed, we are adding a Usage section to show code snippet for major use cases.